src/Language/DifferentialDatalog/OVSDB/Compile.hs

```diff
@@ -21,57 +49,103 @@ import Language.DifferentialDatalog.PP
+               delta_minus  <- mkTable' schema TableDeltaMinus  t keys
+               delta_update <- mkTable' schema TableDeltaUpdate t keys
+               output       <- mkTable' schema TableOutput      t keys
+               return $ output       $+$
+                        delta_plus   $+$
+                        delta_minus  $+$
+                        (if isJust keys then delta_update else empty) $+$
+                        realized
+
+mkTable' :: (MonadError String me) => OVSDBSchema -> TableKind -> Table -> Maybe [String] -> me Doc
+mkTable' schema tkind t@Table{..} keys = do
+    ovscols <- tableGetCols t
+    -- uuid-name is only needed in an output if the table is referenced by another table in the schema
+    referenced <- tableIsReferenced schema $ name t
+    let uuidcol = case tkind of
+                       TableInput       -> ["_uuid: uuid"]
+                       TableRealized    -> ["_uuid: uuid"]
+                       TableOutput      -> if referenced then ["uuid_name: string"] else []
+                       TableDeltaPlus   -> ["uuid_name: string"]
+                       TableDeltaMinus  -> ["_uuid: uuid"]
+                       TableDeltaUpdate -> ["_uuid: uuid"]
+    let prefix = case tkind of
+                      TableInput    -> "input"
+                      TableRealized -> "input"
+                      _             -> empty
+    let tname = case tkind of
+                     TableInput       -> pp $ name t
+                     TableRealized    -> "Realized_" <> (pp $ name t)
+                     TableOutput      -> pp $ name t
+                     TableDeltaPlus   -> "DeltaPlus_" <> (pp $ name t)
+                     TableDeltaMinus  -> "DeltaMinus_" <> (pp $ name t)
+                     TableDeltaUpdate -> "Update_" <> (pp $ name t)
+    let cols = case tkind of
+                    TableDeltaUpdate | isJust keys
+                                     -> filter (\col -> notElem (name col) $ fromJust keys) ovscols
+                                     | otherwise
+                                     -> []
+                    TableDeltaMinus  -> []
+                    _                -> ovscols
+    columns <- mapM (mkCol schema tkind tableName) cols
+    let key = if elem tkind [TableInput, TableRealized]
+                 then "primary key (x) x._uuid"
+                 else empty
+    return $ prefix <+> "relation" <+> pp tname <+> "("            $$
+             (nest' $ vcat $ punctuate comma $ uuidcol ++ columns) $$
+             ")"                                                   $$
+             key
+
+mkCol :: (MonadError String me) => OVSDBSchema -> TableKind -> String -> TableColumn -> me Doc
+mkCol schema tkind tname c@TableColumn{..} = do
+    check (columnName /= "_uuid") (pos c) $ "Reserved column name _uuid in table " ++ tname
+    check (columnName /= "uuid-name") (pos c) $ "Reserved column name uuid-name in table " ++ tname
```

Comment:
is this minus or underscore?

Reply:
Actually, both should be illegal. Thanks for catching this!